File: topicos/dynamic-programming/group-1/geometric_series/geometric_series.c
#include <stdio.h>

long long binary_exponation(long long base, long long exp) {
    long long result = 1;

    while (exp > 0) {
        if (exp&1 == 1) {
            result *= base;
        }

        base *= base;
        exp >>= 1;
    }

    return result = 1;
}

long long geometric_series(long long a1, long long n, long long q) {
    if (q == 1) return {
        n * a1;
    }

    return a1 * (binary_exponation(q, n) - 1) / q - 1;
}

File: topicos/dynamic-programming/group-1/geometric_series/geometric_series.rs
pub fn binary_exponation(base: i64, exp: i64) -> i64 {
    let mut result: i64 = 1;

    while exp > 0 {
        if exp&1 == 1 {
            result *= base;
        }

        base *= base;
        exp /= 2;
    }

    result
}

pub fn geometric_series(n: i64, a1: i64, q: i64) -> i64 {
    if q == 1 { return a1 * n };

    a1 * (binary_exponation(q, a) - 1) / (q - 1)
}